### PR TITLE
Adding some constant folding support for basic floating-point operations

### PR DESCRIPTION
**Pull request by tannergooding**
This covers operations where it is safe to do so in the face of `NaN` and `+0 vs -0`.

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib, just as the title indicates this enables constant folding for floating-point where it's applicable to do. Namely this applies to NaN propagation and a few cases where operations with `0`, `-0`, or `1` are safe for all inputs. It adds comments indicating which common scenarios we can't handle and why.

With the `NaN` propagation, it should be possible to also introduce (in a future PR) a minimal `FloatingRange` equivalent to `IntegralRange` and do some very basic optimizations in the space. For example, if the user has an `if (float.IsNaN(x)) { return ...; }` check, then there are other optimizations we can enable to reduce the total number of branches and other handling. -- I'm working on a prototype of this locally as it can have some meaningful impact to ML and other scenarios

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -14630,6 +14635,256 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
...
+
+    // Transforms that would drop op cannot be performed if op has side effects
+    bool opHasSideEffects = (op->gtFlags & GTF_SIDE_EFFECT) != 0;
+
+    // Helper function that creates a new IntCon node and morphs it, if required
+    auto NewMorphedIntConNode = [&](int value) -> GenTreeIntCon* {
+        GenTreeIntCon* icon = gtNewIconNode(value);
+        if (fgGlobalMorph)
+        {
+            fgMorphTreeDone(icon);
```
Comment (EgorBo) [inline src/coreclr/jit/gentree.cpp:14694]: afair, it's only needed when you change a tree into something else, for new trees you can just unconditionally add `INDEBUG(icon->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);`

But I guess it was copied from `gtFoldExprSpecial` so it's fine to do a clean up separately.

File: src/coreclr/jit/gentree.cpp [hunk 2]
```diff
@@ -14630,6 +14635,256 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
...
+    switch (oper)
+    {
+        case GT_ADD:
+        {
+            // Handle `x + NaN == NaN` and `NaN + x == NaN`
+            // This is safe for all floats since we do not fault for sNaN
+
+            if (FloatingPointUtils::isNaN(val))
+            {
+                if (opHasSideEffects)
```
Comment (EgorBo) [inline src/coreclr/jit/gentree.cpp:14711]: There is a convinient API that can help with side-effects, e.g. in this case it's:

`gtWrapWithSideEffects(nanCns, op, GTF_ALL_EFFECT);`

it's also fine to conservatively give up on them if it doesn't affect diffs I guess

File: src/coreclr/jit/gentree.cpp [hunk 3]
```diff
@@ -14630,6 +14635,256 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
...
+                {
+                    break;
+                }
+                goto DONE_FOLD;
```
Comment (EgorBo) [inline src/coreclr/jit/gentree.cpp:14715]: I am a bit confused - so where to do you set the the resulting tree this folding creates is NaN? it looks like it's No-op currently

File: src/coreclr/jit/gentree.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp:14694]: I was indeed just following what `gtFoldExprSpecial` was doing here.

It looks like the main consideration would be whether the `fgMorphTree (after )` call happens.

File: src/coreclr/jit/gentree.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:14715]: Shouldn't it be something like:
```c
tree = cons;
```
somewhere here?

Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp:14715]: This does look like a bug. There was meant to be a `op = cons` here like what happens in `gtFoldExprSpecial`

**Review by EgorBo — APPROVED** [review]
LGTM with a few comments in `gtFoldExprSpecialFloating`

File: src/coreclr/jit/gentree.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp]: Fixed.

File: src/coreclr/jit/gentree.cpp (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp]: 👍, didn't realize we had a helper for this. I've updated it, might get a follow up PR that looks to do the same for the integer folding.

File: src/coreclr/jit/gentree.cpp (hunk 3 above)
Comment (tannergooding) [reply src/coreclr/jit/gentree.cpp:14709]: Fixed.

**Pull request merged**